src/supervisor.tex

```diff
@@ -2261,3 +2281,108 @@ \chapter{``Svnapot'' Standard Extension for NAPOT Translation Contiguity, Versio
+
+\begin{commentary}
+Future extensions may provide more and/or finer-grained control over which PMAs
+can be overridden.
+\end{commentary}
+
+With Svpbmt enabled, it is possible for multiple virtual aliases of the same
+physical page to exist simultaneously with different memory attributes.
+It is also possible for a U-mode or S-mode mapping through a PTE with Svpbmt
+enabled to observe different memory attributes for a given region of
+physical memory than a concurrent access to the same page performed by M-mode
+or when {\tt satp}.MODE=Bare.  In such cases, each individual access observes
+the memory attributes associated with its own path; aliases are not considered
+when determining attributes.
+
+\begin{commentary}
+For example, a cacheable access may be issued at the same time as a
+non-cacheable access to the same physical memory address.  In this case,
+if the former is performed first in the global memory order, then it will
+be evicted from the cache by the latter.  If on the other hand the cacheable
+access appears after the non-cacheable access, then the former may remain
+cached as it normally would.
+
+Likewise, accesses performed under memory indicating the non-idempotent
+attribute must not be merged with idempotent accesses to the same region
+in flight at the same time, as the non-idempotency of the former must
+be respected.  This is not expected to be a common situation.
+
+Note that Svpbmt cannot be used to completely prevent speculative reads from
+being performed to a region of memory for which the PMAs indicate idempotence,
+as speculation can still be performed via M-mode or via Bare mappings, which do
+not use the PBMTs.
+\end{commentary}
+
+Memory accesses are ordered according to the effective memory type for the
+address in question.  A region of memory which is considered main memory by the
+PMAs but I/O by a PTE will obey channel 0 strong ordering, where for memory
+ordering purposes the ``I/O region'' is considered to be main memory.  In other
+words, each access to such a virtual address will appear in the global memory
+order after all prior operations from the same hart to main memory or to the
+same ``I/O region'', and it will appear earlier in the global memory order than
+all subsequent accesses from the same hart to main memory or to the same ``I/O
+region''.  Such operations are not guaranteed to remain ordered with respect to
+I/O operations to other I/O regions unless FENCEs are inserted by software.
+
+Likewise, a region of memory which is considered I/O by the PMAs but main
+memory by a PTE will obey RVWMO (or RVTSO if enabled) rather than I/O
+strong ordering rules, and accesses to such pages are considered main
+memory rather than I/O for the purposes of FENCE, {\em.aq}, and {\em.rl}.
+In such cases it is the responsibility of software to insert FENCEs
+or other ordering mechanisms if necessary.
+
+\begin{commentary}
+A device driver written to rely on I/O strong ordering rules will not
```

Comment:
This paragraph seems semi-misleading.  While appropriate wrt hyprvisors, an OS is not going to go off on its own and override PMAs, i.e. an OS will apply a pbmt override in response to a request from a device driver.

But I also now realize that this issue can't happen.  If a device driver is relying on I/O strong ordering, then presumably either the PMAs specify "I/O" or the device requested the OS to apply an "I/O" pbmt.  (Otherwise the device driver won't work properly even in a non-virtualized environment.)  In which case, even if the hypervisor maps the address to colloquial "main memory", the address will still be treated as part of an architectural "I/O" region.

Reply:
I somewhat misspoke.  A hypervisor could turn PMA "I/O" into effectively "noncacheable main memory".  But the motivating concern is when a hypervisor maps a guest memory-mapped device into physical main memory for purposes of emulating that device.  This would simply involve G-stage mapping the I/O device's VA pages to physical DRAM main memory.  The G-stage PTE would not override the attributes.  Or if it did, and specifically to "noncacheable main memory",  that would be problematic and unnecessary (i.e. to quote Allen: "if it hurts, then don't do it").

Reply:
Isn't "...then don't do it" exactly what this non-normative text is warning about?  Ultimately either we forbid the combination, or we define what its behavior is.  For now it's not forbidden, so this comment was meant to make the warning explicit rather than leaving it quiet.